### PR TITLE
config (ticdc): fix a data race in changefeed initializing

### LINKED ISSUE #8586: Found a data race during changefeed initializing.
### What did you do?

run integration test case: `charset_gbk`

### What did you expect to see?

no data race.

### What did you see instead?

```
==================
WARNING: DATA RACE
Write at 0x00c006130f30 by goroutine 568:
  github.com/pingcap/tiflow/pkg/config.(*ReplicaConfig).ValidateAndAdjust()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/pkg/config/replica_config.go:214 +0x57c
  github.com/pingcap/tiflow/pkg/config.GetSinkURIAndAdjustConfigWithSinkURI()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/pkg/config/replica_config.go:248 +0xbe
  github.com/pingcap/tiflow/pkg/sink/observer.NewObserver()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/pkg/sink/observer/observer.go:61 +0x1b7
  github.com/pingcap/tiflow/cdc/owner.(*changefeed).initialize()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/owner/changefeed.go:566 +0x1e2d
  github.com/pingcap/tiflow/cdc/owner.(*changefeed).tick()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/owner/changefeed.go:293 +0x2f5
  github.com/pingcap/tiflow/cdc/owner.(*changefeed).Tick()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/owner/changefeed.go:219 +0x395
  github.com/pingcap/tiflow/cdc/owner.(*ownerImpl).Tick()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/owner/owner.go:201 +0xc66
  github.com/pingcap/tiflow/pkg/orchestrator.(*EtcdWorker).Run()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/pkg/orchestrator/etcd_worker.go:290 +0x1b46
  github.com/pingcap/tiflow/cdc/capture.(*captureImpl).runEtcdWorker()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/capture/capture.go:500 +0x309
  github.com/pingcap/tiflow/cdc/capture.(*captureImpl).campaignOwner()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/capture/capture.go:463 +0x1467
  github.com/pingcap/tiflow/cdc/capture.(*captureImpl).run.func2()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/capture/capture.go:328 +0x78
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      /go/pkg/mod/golang.org/x/sync@v0.1.0/errgroup/errgroup.go:75 +0x82

Previous write at 0x00c006130f30 by goroutine 2293:
  github.com/pingcap/tiflow/pkg/config.(*ReplicaConfig).ValidateAndAdjust()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/pkg/config/replica_config.go:214 +0x57c
  github.com/pingcap/tiflow/pkg/config.GetSinkURIAndAdjustConfigWithSinkURI()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/pkg/config/replica_config.go:248 +0xbe
  github.com/pingcap/tiflow/cdc/sink/ddlsink/factory.New()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/sink/ddlsink/factory/factory.go:39 +0xb3
  github.com/pingcap/tiflow/cdc/owner.ddlSinkInitializer()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/owner/ddl_sink.go:121 +0x424
  github.com/pingcap/tiflow/cdc/owner.(*ddlSinkImpl).run.func1()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/owner/ddl_sink.go:154 +0x114

Goroutine 568 (running) created at:
  golang.org/x/sync/errgroup.(*Group).Go()
      /go/pkg/mod/golang.org/x/sync@v0.1.0/errgroup/errgroup.go:72 +0x12e
  github.com/pingcap/tiflow/cdc/capture.(*captureImpl).run()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/capture/capture.go:324 +0x74e
  github.com/pingcap/tiflow/cdc/capture.(*captureImpl).Run()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/capture/capture.go:288 +0x336
  github.com/pingcap/tiflow/cdc/server.(*server).run.func1()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/server/server.go:315 +0x78
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      /go/pkg/mod/golang.org/x/sync@v0.1.0/errgroup/errgroup.go:75 +0x82

Goroutine 2293 (running) created at:
  github.com/pingcap/tiflow/cdc/owner.(*ddlSinkImpl).run()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/owner/ddl_sink.go:150 +0x194
  github.com/pingcap/tiflow/cdc/owner.(*changefeed).initialize()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/owner/changefeed.go:553 +0x1909
  github.com/pingcap/tiflow/cdc/owner.(*changefeed).tick()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/owner/changefeed.go:293 +0x2f5
  github.com/pingcap/tiflow/cdc/owner.(*changefeed).Tick()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/owner/changefeed.go:219 +0x395
  github.com/pingcap/tiflow/cdc/owner.(*ownerImpl).Tick()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/owner/owner.go:201 +0xc66
  github.com/pingcap/tiflow/pkg/orchestrator.(*EtcdWorker).Run()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/pkg/orchestrator/etcd_worker.go:290 +0x1b46
  github.com/pingcap/tiflow/cdc/capture.(*captureImpl).runEtcdWorker()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/capture/capture.go:500 +0x309
  github.com/pingcap/tiflow/cdc/capture.(*captureImpl).campaignOwner()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/capture/capture.go:463 +0x1467
  github.com/pingcap/tiflow/cdc/capture.(*captureImpl).run.func2()
      /home/jenkins/agent/workspace/cdc_ghpr_integration_test/go/src/github.com/pingcap/tiflow/cdc/capture/capture.go:328 +0x78
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      /go/pkg/mod/golang.org/x/sync@v0.1.0/errgroup/errgroup.go:75 +0x82
==================

```

### Versions of the cluster

Upstream TiDB cluster version (execute `SELECT tidb_version();` in a MySQL client):

```console
(paste TiDB cluster version here)
```

Upstream TiKV version (execute `tikv-server --version`):

```console
(paste TiKV version here)
```

TiCDC version (execute `cdc version`):

```console
master
```

### PR DESCRIPTION
**Pull request by asddongmen**
### What problem does this PR solve?

Issue Number: close #8586

#### Tests 

 - Unit test

### Release note 

```release-note
None
```

### CODE REVIEW

File: pkg/config/replica_config.go [hunk 1]
```diff
@@ -240,17 +240,19 @@ func (c *ReplicaConfig) FixMemoryQuota() {
 func GetSinkURIAndAdjustConfigWithSinkURI(
 	sinkURIStr string,
 	config *ReplicaConfig,
-) (*url.URL, error) {
+) (*url.URL, *ReplicaConfig, error) {
```
Comment (amyangfei) [inline pkg/config/replica_config.go:243]: Maybe it is better to make `replicaConfig` to provide some thread-safe API, and separate readonly and writable API. 
`GetSinkURIAndAdjustConfigWithSinkURI` seems mix both read and write in one function.

Comment (asddongmen) [reply pkg/config/replica_config.go]: I changed how to fix this problem. It is better  to not change the value of replicaConfig while changefeed is running. Instead, it should be adjusted in the create or update API.

**Pull request merged**